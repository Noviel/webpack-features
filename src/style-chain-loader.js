import ChainLoader from './chain-loader';
import { isServer } from './target';

export default class StyleChainLoader extends ChainLoader {
  add(name, options, changePropsForEnv) {
    super.add(name, options, changePropsForEnv);

    for (const l of this.loaders) {
      if (typeof l.options.importLoaders === 'number') {
        l.options.importLoaders++;
      }
    }

    return this;
  }
}

function processCSSLoaderForEnv(props, { target, production }) {
  if (isServer(target)) {
    props.name += '/locals';
  }

  return props;
}

export function createCSSLoader(
  { target, production, useStyleLoader = false }, 
  options = {}, 
  cssLoaderProcessor = processCSSLoaderForEnv
) {
  const loader = new StyleChainLoader({ target, production });

  if (useStyleLoader) {
    loader.add('style-loader');
  }

  const cssLoaderOptions = {
    modules: true,
    importLoaders: -1,
    ...options
  };

  loader.add('css-loader', cssLoaderOptions, cssLoaderProcessor);

  return loader;
}
