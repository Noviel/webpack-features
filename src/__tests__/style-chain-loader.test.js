import { createCSSLoader } from '../style-chain-loader';

describe('cssLoader', () => {  
  
  it('should not modify name of css loader for not server env', () => {
    expect(createCSSLoader({ target: 'client', production: true }).get())
      .toMatchSnapshot();
  });

  it('should modify name of css loader for server env', () => {
    expect(createCSSLoader({ target: 'server', production: true }).get())
      .toMatchSnapshot();
  });

  it('should increment every defined importLoaders', () => {
    expect(
      createCSSLoader({ target: 'server', production: true })
        .add('scss')
        .get()
    )
      .toMatchSnapshot();
  });

  it('should add style loader if flag is on', () => {
    expect(
      createCSSLoader({ target: 'client', production: true, useStyleLoader: true })
        .get()
    )
      .toMatchSnapshot();
  });

  it('should use custom options', () => {
    expect(
      createCSSLoader({ target: 'client', production: true, useStyleLoader: true }, {
        modules: false,
        localIdentName: '[hash:base64]',
        minimize: true
      })
        .get()
    )
      .toMatchSnapshot();
  });

});
