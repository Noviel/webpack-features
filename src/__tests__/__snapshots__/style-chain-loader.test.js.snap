// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssLoader should add style loader if flag is on 1`] = `
Array [
  Object {
    "loader": "style-loader",
    "options": Object {},
  },
  Object {
    "loader": "css-loader",
    "options": Object {
      "importLoaders": 0,
      "modules": true,
    },
  },
]
`;

exports[`cssLoader should increment every defined importLoaders 1`] = `
Array [
  Object {
    "loader": "css-loader/locals",
    "options": Object {
      "importLoaders": 1,
      "modules": true,
    },
  },
  Object {
    "loader": "scss",
    "options": Object {},
  },
]
`;

exports[`cssLoader should modify name of css loader for server env 1`] = `
Array [
  Object {
    "loader": "css-loader/locals",
    "options": Object {
      "importLoaders": 0,
      "modules": true,
    },
  },
]
`;

exports[`cssLoader should not modify name of css loader for not server env 1`] = `
Array [
  Object {
    "loader": "css-loader",
    "options": Object {
      "importLoaders": 0,
      "modules": true,
    },
  },
]
`;

exports[`cssLoader should use custom options 1`] = `
Array [
  Object {
    "loader": "style-loader",
    "options": Object {},
  },
  Object {
    "loader": "css-loader",
    "options": Object {
      "importLoaders": 0,
      "localIdentName": "[hash:base64]",
      "minimize": true,
      "modules": false,
    },
  },
]
`;
